ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -0,0 +1,522 @@
+    helpers::{Direction, Error, Role},
+    protocol::{
+        context::{reshard, ShardedContext},
+        prss::{FromRandom, FromRandomU128, SharedRandomness},
+        RecordId,
+    },
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        Sendable, SharedValue,
+    },
+    seq_join::{assert_send, seq_join},
+};
+
+/// This context is only useful for sharded shuffle modules because it implements common operations
+/// that all shards on all helpers perform to achieve the perfect shuffle.
+trait ShuffleContext: ShardedContext {
+    /// This sends a single machine word (8 byte value) to one of the helpers specified in
+    /// `direction` parameter.
+    fn send_word(
+        self,
+        direction: Direction,
+        val: usize,
+    ) -> impl Future<Output = Result<(), crate::error::Error>> + Send {
+        async move {
+            Ok(self
+                .set_total_records(1)
+                .send_channel::<BA64>(self.role().peer(direction))
+                .send(
+                    RecordId::FIRST,
+                    BA64::truncate_from(u128::try_from(val).unwrap()),
+                )
+                .await?)
+        }
+    }
+
+    /// This receives a single machine word (8 byte value) to one of the helpers specified in
+    /// `direction` parameter.
+    fn recv_word(
+        self,
+        direction: Direction,
+    ) -> impl Future<Output = Result<usize, crate::error::Error>> + Send {
+        async move {
+            let val: [u8; 8] = self
+                .recv_channel::<BA64>(self.role().peer(direction))
+                .receive(RecordId::FIRST)
+                .await?
+                .as_raw_slice()
+                .try_into()
+                .unwrap();
+            Ok(usize::try_from(u64::from_le_bytes(val)).unwrap())
+        }
+    }
+
+    /// In sharded shuffle, it is a common operation to apply a mask to the input and then permute
+    /// it, with all shards involved in the process. This routine implements it, so helper-specific
+    /// code does not need to do this repetitive work.
+    ///
+    /// The destination shard for each masked row is decided based on value obtained from sampling
+    /// PRSS. Which value to use (left or right) is decided based on `direction` parameter.
+    fn mask_and_shuffle<I, S>(
+        self,
+        direction: Direction,
+        data: I,
+    ) -> impl Future<Output = Result<Vec<S>, crate::error::Error>> + Send
+    where
+        I: IntoIterator<Item = S>,
+        I::IntoIter: ExactSizeIterator + Send,
+        S: ShuffleShare,
+    {
+        let data = data.into_iter();
+        async move {
+            let masking_ctx = self.narrow(&ShuffleStep::Mask);
+            let mut resharded = assert_send(reshard(
+                self.clone(),
+                data.enumerate().map(|(i, item)| {
+                    let (l, r) = masking_ctx.prss().generate(RecordId::from(i));
```

Comment:
One thing I like about including the shared identifier in the AES input is that we can explicitly see that they are different. Otherwise there is an external dependency on another system that's not captured by the logic of the code.

Reply:
It's a fair point, when we discussed the IPA protocol back in Feb, we couldn't find a case where we needed all shards to generate the same value from PRSS, so we made all PRSS shards seeded with unique AES keys inside in-memory infrastructure: https://github.com/private-attribution/ipa/blob/8808dc5bd3bdb682229d727433008089aa6647f6/ipa-core/src/test_fixture/world.rs#L690

I'd be inclined to let protocols rely on Infrastructure initializing PRSS correctly, similarly to how it currently trusts PRSS negotiation to run DH correctly and set up unique shared keys per helper pair. I agree that it would be great to have some tests that check this behavior. It would be great if we can collect all PRSS values generated on each shard, write them to a file and after query completes, check for duplicates.

Reply:
This makes sense to me. This is something we should document in the IETF.